util/session/sessionmanager.go

```diff
@@ -490,86 +494,155 @@ type TokenVerifier interface {
+			log.Warnf("Failed to get OIDC provider for JWT verification: %v", provErr)
+		} else {
+			token, jwtErr := prov.VerifyJWT(tokenString, argoSettings)
+			if jwtErr == nil {
+				// Successfully verified as JWT via JWKS URL
+				log.Debug("Token verified using JWT config (JWKS URL)")
+				return token.Claims, "", nil
+			}
+			// Log the JWT verification failure and continue to other methods
+			log.Debugf("JWT verification failed, trying other methods: %v", jwtErr)
 		}
+	}
 
-		argoSettings, err := mgr.settingsMgr.GetSettings()
-		if err != nil {
-			return nil, "", fmt.Errorf("cannot access settings while verifying the token: %w", err)
-		}
-		if argoSettings == nil {
-			return nil, "", errors.New("settings are not available while verifying the token")
-		}
+	// 2. Parse token unverified to check issuer for Argo CD or OIDC flow
+	parser := jwt.NewParser(jwt.WithoutClaimsValidation())
+	claims := jwt.MapClaims{}
+	_, _, parseErr := parser.ParseUnverified(tokenString, &claims)
+	if parseErr != nil {
+		// If we couldn't even parse it, and JWT verification didn't work (or wasn't configured), fail.
+		return nil, "", fmt.Errorf("failed to parse token: %w", parseErr)
+	}
+
+	// 3. Check if it's an Argo CD issued token
+	issuer, issErr := claims.GetIssuer()
+	if issErr != nil {
+		log.Debugf("Could not get issuer claim, assuming not Argo CD token: %v", issErr)
+	} else if issuer == SessionManagerClaimsIssuer {
+		log.Debug("Attempting verification as Argo CD token")
+		return mgr.Parse(tokenString)
+	}
 
-		idToken, err := prov.Verify(tokenString, argoSettings)
-		// The token verification has failed. If the token has expired, we will
-		// return a dummy claims only containing a value for the issuer, so the
-		// UI can handle expired tokens appropriately.
-		if err != nil {
-			log.Warnf("Failed to verify token: %s", err)
-			tokenExpiredError := &oidc.TokenExpiredError{}
-			if errors.As(err, &tokenExpiredError) {
-				claims = jwt.MapClaims{
-					"iss": "sso",
-				}
-				return claims, "", common.ErrTokenVerification
+	// 4. Attempt OIDC verification (external IDP or Dex)
+	log.Debugf("Attempting verification as OIDC token (issuer: %s)", issuer) // Use issuer variable from above
+	prov, provErr := mgr.provider()
+	if provErr != nil {
+		// If OIDC/Dex is not configured, but we reached here, the token is invalid
+		return nil, "", fmt.Errorf("token issuer (%s) is not '%s' and OIDC provider is not available: %w", issuer, SessionManagerClaimsIssuer, provErr)
+	}
+
+	idToken, oidcErr := prov.Verify(tokenString, argoSettings)
+	if oidcErr != nil {
+		log.Warnf("OIDC token verification failed: %s", oidcErr)
+		// Handle expired token specifically for UI hints using errors.As
+		var tokenExpiredError *oidc.TokenExpiredError
+		if errors.As(oidcErr, &tokenExpiredError) {
+			// Return minimal claims indicating SSO source for expired tokens
+			// Use issuer variable from above, handle potential error if it wasn't retrieved
+			issForExpired := ""
```

Comment:
default to sso if not found?
```suggestion
			issForExpired := "sso"
```